Update bytesize-6-dsl2-using-modules-pipelines.md (#731)
nf-core/nf-co.re: markdown/events/2021/bytesize-6-dsl2-using-modules-pipelines.md

Thank you very much and welcome to the bytesize talk on how to use modules in a pipeline.

I would just like to briefly recap what a module is. It is an atomic process that cannot be reduced any further and usually contains a single software tool like FastQC for example and it can be used within a pipeline and also shared between different pipelines. To make use of the sharing, there is an nf-core modules repository on GitHub where you can find many of these modules already.

So to make use of this modules repository, there’s a new sub-command in nf-core/tools and for a brief recap, you can install it with ´pip install nf-core´ and ´conda install nf-core´ and if you then run ´nf-core modules´, you get a list of sub-commands that you can use to interact with this repository.

I would just like to briefly recap what a module is. It is an atomic process that cannot be reduced any further and usually contains a single software tool like FastQC for example and it can be used within a pipeline and also shared between different pipelines. To make use of the sharing, there is an nf-core modules repository on GitHub where you can find many of these modules already.

> With one of the sub-commands, and with --help, you get some instructions on how to use them, so over the next couple of slides, I will briefly introduce you to some of the sub-commands that could be helpful for using modules in the pipeline. [0:47](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=51)

Reviewer: I think it would make sense to put the time stamps at the begininig of the lines, I think one of the other bytesize transcriptions did this? At least I see it more often that way around in other contexts (but this is just a suggestion)

Reviewer: I see you've done this further down so be consistent one way or th eother and it's fine!

Author: Thanks. 
I've tried to have them at the start of the lines, but it makes sense to have them in the middle of sentences in certain places as well because it can directly relate to what's been said at that moment. 
I'll revise this where appropriate.

One of the first things you might want to do is to try and find a module that you could use, and you can use the ´nf-core modules list´ for that. [1:30](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=90)

> That will just print out all the modules currently available and they are subdivided into tool and sub-tool because many tools like ´sam tools´ or ´bcf tools´ have these sub-commands that are then their own module basically.

Reviewer: ```suggestion
That will just print out all the modules currently available and they are subdivided into tool and sub-tool because many tools like ´samtools´ or ´bcftools´ have these sub-commands that are then their own module basically.
```

It then looks like this, so on the left hand side, there is a completely fresh new pipeline that I created with the template and then I ran ‘nf-core modules install´and in the green part you can see where this module ends up. [2:12](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=132)

The module basically creates a subfolder called nf-core software where you can find the FASTQC module, and with the three files i.e. the ´functions.nf´, ´main.nf´ and ´meta.yml´. In the meta.yml, you can find documentation such as what type of input and output this module takes, who wrote it and these sort of things.

In the main nf, that’s where the extra magic happens, where the fastqc is run, and in the ´functions.nf´ there are some helper functions that are needed.

[3:03](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=183) I guess the opposing step to that is how to remove a module that you no longer want to use or that you erroneously installed.

For that you run ´nf core modules remove´ and that removes this entire subfolder that had fastqc with the functions main and meta file in it. You don’t have to do anything else there.

> [3:25](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=206) Then sometimes the module will get updated, maybe the software was updated and this was propagated already to the module, and you would like to use thus update. So how do we ensure that first of all we have the newest version, and then how to update it?

Reviewer: ```suggestion
[3:25](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=206) Then sometimes the module will get updated, maybe the software was updated and this was propagated already to the module, and you would like to use this update. So how do we ensure that first of all we have the newest version, and then how to update it?
```

> So to check this module, you can run ´nf core modules lint´ on the directory to check all modules, or on a specific module like FASTQC.

Reviewer: ```suggestion
So to check this module, you can run ´nf core modules lint´ on the directory to check all modules, or on a specific module like fastqc.
```

It will check among other things whether or not you have missed any changes. Then to update it, you currently have to remove it and reinstall it.

But for the future, Kevin is working on creating an ´update´ subcommand that you can then use.

[4:13](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=253) OK, so last but not least, what do you do if the software you’re looking for in not in nf-core/modules?

You basically have two options, the first is to add it to nf-core/modules, and the second is to create a local module. So to add it to nf-core/modules, this is usually really helpful if others use this pipeline or software as well.

However, if you’re unsure about this, you can always check the issues to see if someone else has already started working on it. Alternatively, ask in the #modules channel in Slack. Kevin will also talk about how to do this in his presentation.

[4:57](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=297) The other option is to create a local module, so this is useful for software that is specific for your pipeline.

You can run ´nf core modules create´, which will create a local module for you.

> If you look in this box, you can see the mf-core/modules folder that we’ve already seen before, and the local subfolder in which your tools will then live. A lot of the things that Harshil and Kevin covered in their talks is relevant here, and Kevin’s talk is more tailored towards modules, but a lot of the functionality will be similar there.

Reviewer: ```suggestion
If you look in this box, you can see the nf-core/modules folder that we’ve already seen before, and the local subfolder in which your tools will then live. A lot of the things that Harshil and Kevin covered in their talks is relevant here, and Kevin’s talk is more tailored towards modules, but a lot of the functionality will be similar there.
```

[5:36](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=336) OK, so this ´nf core modules create´ gives you this file here with many to-do statements and little help messages, and you can start filling out and try to get your tool to run here.

Hopefully the to-do statements help you figure out what exactly you will need to do in each step.

> [5:58](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=358) So now that we have modules either local or nf-core/modules, we want to start writing actual workflows and pipelines.

Reviewer: ```suggestion
[5:58](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=358) So now that we have either local or nf-core/modules, we want to start writing actual workflows and pipelines.
```

[9:46](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=586) So last but not least, we have to combine the sub-workflows to workflows or sub-workflows of modules to workflows. So in the header, we have the two ´include’ statements for the workflows or the sub-workflows and the ‘fastqc’, i.e. once from the modules and the nf-core software, and one from the sub-workflows local.

Once again we have to add the parameters to parse them down and I have highlighted that in yellow here [10:11](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=613) for you to be able to track this.

For the FASTQC module, we just need to specify the parameters for fastqc, so modules [´fastqc´], but for the sub-workflow one needs to specify all options for all the tools in there.

[10:34](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=634) But for the sub-workflow, you need to specify all options for the tools. So here we have ´bwa_mem´ or ´samtools_sort´, and this is exactly now this field that was originally specified in ´conf/modules.config´.

> [10:47](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=645) OK, then in the workflow and in nf here I have just run the module ´FASTQC´ on my input reads that I got and then the sub-workflow on the reads and the index, I get the ´sorted.bams´ output, and then I can do some more steps with that.

Reviewer: ```suggestion
[10:47](https://youtu.be/tWvou0xj9wA?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=645) OK, then in the workflow and in nf here I have just run the module FASTQC on my input reads that I got and then the sub-workflow on the reads and the index, I get the sorted.bams output, and then I can do some more steps with that.
```

**Note: This text has been edited to make it more suitable for reading.**

#### DSL2 - Using modules in a pipeline

Thank you very much and welcome to the bytesize talk on how to use modules in a pipeline.

I would just like to briefly recap what a module is. It is an atomic process that cannot be reduced any further and usually contains a single software tool like FastQC for example and it can be used within a pipeline and also shared between different pipelines. To make use of the sharing, there is an nf-core modules repository on GitHub where you can find many of these modules already.

> So to make use of this modules repository, there’s a new sub-command in nf-core/tools and for a brief recap, you can install it with ´pip install nf-core´ and ´conda install nf-core´ and if you then run ´nf-core modules´, you get a list of sub-commands that you can use to interact with this repository.

Reviewer: ```suggestion
So to make use of this modules repository, there’s a new sub-command in nf-core/tools and for a brief recap, you can install it with `pip install nf-core` and `conda install nf-core` and if you then run `nf-core modules`, you get a list of sub-commands that you can use to interact with this repository.
```

Reviewer: might as well do this replacement everywhere rather than the odd single quotes here as then you code blocks which is basically what these are

Author: Thanks. Good to learn how to use the quote marks for code blocks. :-) Thought they were the other way around.

Author: Changed them throughout the text. Thanks :-)

Reviewer: I always use this page when I forget various Markdown things :)

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code

> I would just like to briefly recap what a module is. It is an atomic process that cannot be reduced any further and usually contains a single software tool like FastQC for example and it can be used within a pipeline and also shared between different pipelines. To make use of the sharing, there is an nf-core modules repository on GitHub where you can find many of these modules already.

Reviewer: Copy paste error from first para?

Author: Deleted!
